soda/contracts/README.md

```diff
@@ -0,0 +1,107 @@
+
+checks:
+  - avg(distance) between 400 and 500
+```
+
+# Top level keys
+
+| Key | Description | Required | YAML data type |
+| --- | ----------- | -------- | -------------- |
+| `name` | Name of the dataset as in the SQL engine. | Required | string |
+| `columns` | Schema specified as a list of columns.  See below for the column format | Required| list of objects |
+| `checks` | SodaCL checks.  See [SodaCL docs](https://docs.soda.io/soda-cl/metrics-and-checks.html) | Optional | list of SodaCL checks |
+
+# Column keys
+
+| Key | Description | Required | YAML data type |
+| --- | ----------- | -------- | -------------- |
+| `name` | Name of the column as in the SQL engine. | Required | string |
+| `data_type` | Ensures verification of this physical data type as part of the schema check. Must be the name of the data type as in the SQL engine | Optional | string |
+| `not_null` | `not_null: true` Ensures a missing values check | Optional | boolean |
+| `missing_*` | Ensures a missing values check and uses all the `missing_*` configurations | Optional | list of strings or numbers |
+| `valid_*` * `invalid_*` | Ensures a validity check and uses all the `valid_*` and `invalid_*` configurations | Optional | list of strings or numbers |
+| `unique` | `unique: true` ensures a uniqueness check | Optional | boolean |
+
+# Contract to SodaCL translation
+
+## The schema check
+
+A contract verification will always check the schema.  The contract schema check will verify that the list of columns
+in the database dataset matches *exactly* with the columns in the contract file.  All columns listed in the contract
+are required and no other columns are allowed.
+
+Optionally, if the `data_type` property is specified in the column, the data type will be checked as well as part of
+the schema check.
+
+## Other column checks
+
+`not_null`, `missing_values`, `missing_format` or `missing_regex` will ensure a single check
+`missing_count({COLUMN_NAME}) = 0` on that column with all the `missing_*` keys as configuration.
+
+Presence of `valid_*`, `invalid_*` column keys will ensure a single validity check `invalid_count({COLUMN_NAME}) = 0`
+on that column with all `valid_*`, `invalid_*` keys as configuration.
+
+`unique: true` will ensure a SodaCL check `duplicate_count({COLUMN_NAME}) = 0`
+
+## SodaCL checks section
+
+The YAML list structure under `checks:` will just be copied to the SodaCL checks file as-is.
+
+# Contract enforcement
+
+"Enforcement" of a contract comes down to verifying that a certain dataset (like eg a table) complies with the specification in
+the contract file.  When the contract does not comply, the data owner and potentially the consumers should be notified.
+
+In your python (virtual) environment, ensure that the libraries `soda-core` and `soda-core-contracts` are available
+as well as the `soda-core-xxxx` library for the SQL engine of your choice.
+
+To verify if a dataset complies with the contract, here's the code snippet.
+
+```python
```

Comment:
is is the case that we won't offer a CLI option to enforce a contract?

Reply:
I guess it depends how much in the future we're talking about. IMO the programmatic way is shitty to maintain cos you have to reapeat all this boilerplate code which, unless you have a good reason to go for programmatic I think most users would choose CLI.

To me it also seems rather trivial to add a new command or option to the scan command to also do the translation step. But maybe I don't see something